python/ray/dag/compiled_dag_node.py

```diff
@@ -1864,11 +1909,62 @@ def _check_inputs(self, args: Tuple[Any, ...], kwargs: Dict[str, Any]) -> None:
                     f"must be called with kwarg `{kwarg}`"
                 )
 
+    def _has_execution_results(
+        self,
+        execution_index: int,
+    ) -> bool:
+        """Check whether there are results corresponding to the given execution
+        index stored in self._result_buffer. This helps avoid fetching and
+        storing results again.
+
+        Args:
+            execution_index: The execution index corresponding to the result.
+
+        Returns:
+            Whether the result for the given index has been fetched and cached.
+        """
+        return execution_index in self._result_buffer
+
+    def _cache_execution_results(
+        self,
+        execution_index: int,
+        channel_index: Optional[int],
+        result: Optional[Any] = None,
```

Comment:
I would split this into a cache setter and getter method to make this logic more robust, instead of allowing `result` to be optional.

Ideally, `_execute_until` should also use these methods.